pysat/instruments/madrigal_pandas.py

```diff
@@ -0,0 +1,187 @@
+    
+    code = self.kwargs['madrigal_code']
+    tag = self.kwargs['madrigal_tag']
+    self._download_rtn = functools.partial(_general_download, 
+                                      inst_code=str(code),
+                                      kindat=str(tag))
+    return
+
+def _general_download(date_array, tag='', sat_id='', data_path=None, user=None,
+             password=None, inst_code=None, kindat=None):
+    """Downloads data from Madrigal.
+    
+    Method will be partially set using functools.partial. Intended to
+    have the same call structure as normal instrument download routine.
+    Upon Instrument instantiation this routine will be set to
+    parameters specific to a Madrigal data set. It will then work like
+    a standard download call.
+    
+    Parameters
+    ----------
+    date_array : array-like
+        list of datetimes to download data for. The sequence of dates need not
+        be contiguous.
+    tag : string ('')
+        Tag identifier used for particular dataset. This input is provided by
+        pysat.
+    sat_id : string  ('')
+        Satellite ID string identifier used for particular dataset. This input
+        is provided by pysat.
+    data_path : string (None)
+        Path to directory to download data to.
+    user : string (None)
+        User string input used for download. Provided by user and passed via
+        pysat. If an account
+        is required for dowloads this routine here must error if user not
+        supplied.
+    password : string (None)
+        Password for data download.
+    inst_code : int
+        Madrigal integer code used to identify platform
+    kindat : int
+        Madrigal integer code used to identify data set
+        
+    Returns
+    --------
+    Void : (NoneType)
+        Downloads data to disk.
+    
+    Notes
+    -----
+    The user's names should be provided in field user. Ruby Payne-Scott should
+    be entered as Ruby+Payne-Scott
+
+    The password field should be the user's email address. These parameters
+    are passed to Madrigal when downloading.
+
+    The affiliation field is set to pysat to enable tracking of pysat downloads.
+
+    """
+    mad_meth.download(date_array, inst_code=inst_code,
+                      kindat=kindat,
+                      data_path=data_path, user=user, password=password)
+
+   
+        
+def clean(self):
+    """Routine to return JRO ISR data cleaned to the specified level
+
+    Returns
+    --------
+    Void : (NoneType)
+        data in inst is modified in-place.
+
+    Notes
+    --------
+    Supports 'clean', 'dusty', 'dirty'
+    'Clean' is unknown for oblique modes, over 200 km for drifts
+    'Dusty' is unknown for oblique modes, over 200 km for drifts
+    'Dirty' is unknown for oblique modes, over 200 km for drifts
+    'None' None
+
+    Routine is called by pysat, and not by the end user directly.
+    
+    """
+    if self.clean_level in ['clean', 'dusty']:
```

Comment:
Should also include "dirty" in this list.

Reply:
Updated